attestation.go

```diff
@@ -62,30 +61,72 @@ var (
+//
+// Verify that Attested Certificate is signed by that Attestation Certificate
+//
+// The `attestationCert` is the Certificate from the Yubikey Attestation slot,
+// signed by the provided roots. It should not be a CA Certificate in most cases.
+//
+// The `attestedCert` is the Certificate signed by the Attestation slot asserting
+// that the public key was generated on-chip.
+//
+// The `options` is the set of roots and verification assertions to use when checking
+// the `attestationCert` and `attestedCert`. If the `attestationCert` is not a CA, then
+// the path length of roots given should be increased by one, so that the `attestationCert`
+// can be used as a CA to verify the `attestedCert` back to a trusted root
+func VerifyAttestationWithOptions(attestationCert, attestedCert *x509.Certificate, options x509.VerifyOptions) (verifiedChains [][]*x509.Certificate, err error) {
+	// Verify Attestation Cert against CA
+	if _, err = attestationCert.Verify(options); err != nil {
+		return
+	}
+
+	// Initialize intermediate chains if necessary
+	if options.Intermediates == nil {
+		options.Intermediates = x509.NewCertPool()
+	}
+
+	// Add Attestation Certificate to Trust Store
+	// Since this likely is not and probably should not be a CA we pretend it is
+	// to verify the attested cert to the CAs given in the options
+	if !attestationCert.IsCA {
+		attestationCert.IsCA = true
+		attestationCert.BasicConstraintsValid = true
+		attestationCert.MaxPathLen = 1
+		attestationCert.KeyUsage |= x509.KeyUsageCertSign
+	}
+	options.Intermediates.AddCert(attestationCert)
+
+	return attestedCert.Verify(options)
+}
+
+// Verify that the attestation certificate is correctly signed by the Yubikey
+// root (provided with this package), as well as verifying that the Attestation
 // Certificate is signed by that slot correctly.
 //
-// The `attestationCert` is the Certificate from the Yubikey Attestion slot,
-// signed by the Yubico roots.
+// The `attestationCert` is the Certificate from the Yubikey Attestation slot, signed by
+// the Yubico roots. It is not a CA certificate from Yubico.
 //
-// The `attestedCert` is the Certificate signed by the Attestion slot asserting
+// The `attestedCert` is the Certificate signed by the Attestation slot asserting
 // that the public key was generated on-chip.
 func VerifyAttestation(attestationCert, attestedCert *x509.Certificate) ([][]*x509.Certificate, error) {
 	roots := x509.NewCertPool()
-	if !roots.AppendCertsFromPEM(yubicoAttestionRoot2018) {
-		return nil, fmt.Errorf("ykpiv: INTERNAL ERROR: Attestion Root PEM is wrong!")
+	for _, yubicoPivAttestationCA := range yubicoPivAttestationCAs {
+		block, _ := pem.Decode(yubicoPivAttestationCA)
+		caCert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			return nil, fmt.Errorf("ykpiv: INTERNAL ERROR: Attestion Root PEM is wrong!")
+		}
+		// Extend the path length so we can pretend that attestation cert is also a CA
+		caCert.MaxPathLen += 1
```

Comment:
NB that the old CA cert has an incorrect path len of 0, so you'll need to explicitly set it to '1' before if the value is 0 and then incrementing here.
```
openssl x509 -in piv-attestation-ca.pem -text -noout
[snip]
        X509v3 extensions:
            X509v3 Subject Key Identifier: 
                CA:5F:CA:F2:C4:A2:31:9C:E9:22:5F:F1:EC:F4:D5:DF:02:BF:83:BF
            X509v3 Basic Constraints: 
                CA:TRUE, pathlen:1
            X509v3 Key Usage: critical
                Certificate Sign, CRL Sign
```
vs.
```
openssl x509 -in piv-attestation-ca-old.pem -text -noout
[snip]
        X509v3 extensions:
            X509v3 Subject Key Identifier: 
                CA:5F:CA:F2:C4:A2:31:9C:E9:22:5F:F1:EC:F4:D5:DF:02:BF:83:BF
            X509v3 Basic Constraints: 
                CA:TRUE, pathlen:0
            X509v3 Key Usage: critical
                Certificate Sign, CRL Sign
```

Reply:
So I now just realized `attestationCert.MaxPathLen = 1` should be `attestationCert.MaxPathLen = 0` since the attested certificate will never be a CA. It should not be necessary to check if pathLength = 0 on a CA since it can still issue leaf certificates, just no intermediate CAs.

The old and the new CA are the same CA, just that the new PEM allows for an intermediate CA to be issued off the root. It should not be necessary to add the old PEM.

Reply:
Good catch. It would be helpful to run this through some test cases, but any leaf attestation certs will have a Yubikey serial in them, and you or other contributors may not want to include those in tests since that could de-anonymize the contributor where those keys have been used for other purposes.

Would be nice if Yubico could supply some test certs.